### ISSUE TITLE
IATI organisation file support

### ISSUE THREAD

**Issue by siemvaessen** [OpenDataServices/cove#959]
We are looking into CoVe for converting IATI organisation file types, next to the activity file. 

We don't see support for this in CoVe (yet?) so we are wondering what your thinking is on implementing such a feature and if you do not have short term plans to do so, we'd be happy to talk about how we could contribute to this.

**Comment by Bjwebb** [comment]
Yes, support for this doesn't exist yet, and it makes sense to implement this in CoVE.
Our conversion library flatten-tool nearly supports this, it's just missing a way to name the root XML element (`iati-organisations`). There's an issue for this https://github.com/OpenDataServices/flatten-tool/issues/150

There's also some discussion in this issue - https://github.com/OpenDataServices/cove/issues/732 - about how CoVE might auto detect organisation spreadsheets from the Meta tab.

We have no short terms plans to do anything on this, so would by happy to see you contribute.

**Comment by siemvaessen** [comment]
Thanks @Bjwebb. On the library flatten-tool, any ETA on delivery? 

And, where should we look for relevant code in CoVe on the org file detection yet?

**Comment by Bjwebb** [comment]
Sorry for the delay on this, I had a long weekend away at FOSDEM.

We don't have an ETA on that flatten-tool issue atm, but it should be fairly straightforward, so something we can slot into our backlog. How quickly are you hoping to get this functionality?

For the CoVE the changes need to happen in a few places. I'm thinking a call might be useful to talk over these?

**Comment by siemvaessen** [comment]
Hi @Bjwebb - delay in answering. Will CoVE support this in Q2 this year? We'd be happy to contibute, but I'd first like to find out where CoVE now stands with org.file support. We could on the base what is out there continue building.

**Comment by Bjwebb** [comment]
We don't currently have this on our roadmap, but we'd definitely be happy to support you to do this.

**Comment by siemvaessen** [comment]
check. Closing this ticket then.

**Comment by andylolz** [comment]
Any chance of reopening this issue? I’d also find org file support useful.

**Comment by andylolz** [comment]
Thanks for reopening, @Bjwebb!

> we'd definitely be happy to support you to do this.

I don’t know how far this invitation extends?! But OpenDataServices/flatten-tool#201 is my first blocker. Can I get a sense of whether / when it will be reviewed / merged? Thanks!

**Comment by Bjwebb** [comment]
Thanks, I'll try and take a look at that this week.

**Comment by siemvaessen** [comment]
Bumping this up. @Bjwebb any updates on this? We'd be happy to assist, seeing we need it :-) so let me know, also happy to setup a brief call for specs so we can continue.

**Comment by andylolz** [comment]
I started making progress on this (with OpenDataServices/flatten-tool#201). It turns out a bunch of things need to happen first. Next up on my list is: OpenDataServices/flatten-tool#176

**Comment by siemvaessen** [comment]
@andylolz - how can we contribute. Got a plan of action so we can distribute tasks?

So PR Nr. 201 has been accepted, so it's just #180 ?

**Comment by andylolz** [comment]
@siemvaessen really good point. I know back in April I had a plan for this, and I regret not detailing it in a ticket (in fact I can see I did a bunch of unstaged work on this, too.)

I’ll try to retrace my steps, and will break it down here next week.

**Comment by Bjwebb** [comment]
I've not had chance to look into what's still needed for organisation files in CoVE. @andylolz if you do write  up a plan, that would be helpful. I'm happy to check it.

**Comment by andylolz** [comment]
> if you do write up a plan, that would be helpful.

Sure, why not.

> I'm happy to check it.

Thanks!

**Comment by Bjwebb** [comment]
I will be doing some work on IATI Organisation file support this week.

**Comment by andylolz** [comment]
> I will be doing some work on IATI Organisation file support this week.

Haha, kk. So as mentioned in https://github.com/OpenDataServices/cove/pull/993#issuecomment-381556434, I think this was blocked by OpenDataServices/flatten-tool#196, which is now sorted. But by all means.

**Comment by Bjwebb** [comment]
Here's a dev deploy of the work so far: http://959-org-file.dev.cove.opendataservices.coop/
Here's a spreadsheet file to test with it: https://github.com/OpenDataServices/cove/raw/959-org-file/cove_iati/fixtures/basic_iati_org_valid.xlsx

What works:
* Spreadsheet to valid IATI organisation XML (including correct sorting) for simple data
* CoVE validates against the IATI organisation schema
* Here's a list of all possible headings https://gist.github.com/Bjwebb/e32f5139e62ce78ba923cdfcbfdf6cbd generated by this code https://github.com/OpenDataServices/flatten-tool/pull/215

Still TODO:
- [ ] One to many relationships, because we need to use `organisation-identifier` as the key instead of `iati-identifier`
- [ ] CoVE doesn't tell you that you've uploaded an organisation file, or that it is validating against the organisation schema
- [ ] More work on tests
- [ ] Multilingual template generation

Additionally IATI organisation XML -> spreadsheet doesn't work, you just get a blank spreadsheet. I am unlikely to get time to fix this.

**Comment by andylolz** [comment]
COOL!

> Additionally IATI organisation XML -> spreadsheet doesn't work, you just get a blank spreadsheet. I am unlikely to get time to fix this.

^^ Right yeah, I guess this bit is about meta tab roundtripping.

**Comment by Bjwebb** [comment]
Okay, so I think we're done in terms of CoVE feature development for this PR. Still some to do work on tests and template generation.

Many thanks to @andylolz for having done a lot of the work on this!

Dev deploy: http://959-org-file.dev.cove.opendataservices.coop/
Here's a spreadsheet file to test with it, that's been updated to have multiple tabs with one to many relationships: https://github.com/OpenDataServices/cove/raw/959-org-file/cove_iati/fixtures/basic_iati_org_valid.xlsx

**Comment by Lathrisk** [comment]
https://github.com/OpenDataServices/cove/issues/959#issuecomment-413179663

Organisation XML -> Spreadsheet is now implemented (pending deployment), but round trip is still not possible as the spreadsheet is still missing the Meta-tab. 

Multilingual support is still missing also.

See: https://github.com/OpenDataServices/cove/pull/1219

**Comment by stevieflow** [comment]
What's blocking us getting this to deployed @Bjwebb @Lathrisk ?

**Comment by Lathrisk** [comment]
@stevieflow I'm just updating the PR and refreshing my memory just now, but it will need re-reviewed (@Bjwebb had previously requested changes). I'm not sure if the meta-tab work that Ben has been doing will tie in with this as well? I will mention it at stand-up tomorrow morning.
